### PR TITLE
Fix for JSON on optional nested types.

### PR DESCRIPTION
**Pull request by kschwab**
Resolves https://github.com/pydantic/pydantic-settings/issues/196

Updated to recursively search the annotation for any model that matches key. Note that there is a possibility that it could get the wrong field in case of union with complex types that both have the same key. Not sure how that case should be handled. For now, it picks first match.

### CODE REVIEW

**Comment by hramezani** [comment]
Thanks @kschwab for this PR 🙏 

The test case that you've added is great. Could you please add another test case based on the code snippet in reported issue?

**Comment by kschwab** [comment]
@hramezani good catch. #196 was slightly different issue compared to mine. I have pushed a fix and added test.

File: tests/test_settings.py [hunk 1]
```diff
@@ -1212,6 +1228,19 @@ class Settings(BaseSettings):
...
+    class Foo(BaseModel):
+        a: int
+
+    class Settings(BaseSettings):
+        foo: Optional[Foo] = None
+
+    p = tmp_path / 'foo'
+    p.write_text('{"a": 10}')
+
+    assert Settings(_secrets_dir=tmp_path).model_dump() == {'foo': {'a': 10}}
```
Comment (hramezani) [inline tests/test_settings.py:1241]:
```suggestion
    p = tmp_path / 'foo'
    p.write_text('{"a": 10}')

    class Foo(BaseModel):
        a: int

    class Settings(BaseSettings):
        foo: Optional[Foo] = None

        model_config = SettingsConfigDict(secrets_dir=tmp_path)

    assert Settings().model_dump() == {'foo': {'a': 10}}
```

**Review by hramezani — APPROVED** [review]

**Comment by hramezani** [comment]
Thanks @kschwab for this. I've just updated the tests.

**Comment by kschwab** [comment]
Thanks @hramezani!

**Pull request merged**